src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1800,20 +1812,32 @@ public ImmutableArray<DocumentId> GetDocumentIdsWithFilePath(string? filePath)
         private ImmutableDictionary<ProjectId, CompilationTracker> CreateCompilationTrackerMap(ProjectId projectId, ProjectDependencyGraph dependencyGraph)
         {
             var builder = ImmutableDictionary.CreateBuilder<ProjectId, CompilationTracker>();
-            var dependencies = dependencyGraph.GetProjectsThatTransitivelyDependOnThisProject(projectId);
+            IEnumerable<ProjectId>? dependencies = null;
 
-            foreach (var projectIdAndTracker in _projectIdToTrackerMap)
+            foreach (var (id, tracker) in _projectIdToTrackerMap)
             {
-                var id = projectIdAndTracker.Key;
-                var tracker = projectIdAndTracker.Value;
-
                 if (!tracker.HasCompilation)
                 {
                     continue;
                 }
 
-                var canReuse = id == projectId || !dependencies.Contains(id);
-                builder.Add(id, canReuse ? tracker : tracker.Fork(tracker.ProjectState));
+                builder.Add(id, CanReuse() ? tracker : tracker.Fork(tracker.ProjectState));
+
+                // Returns true if 'tracker' can be reused for project 'id'
+                bool CanReuse()
+                {
+                    if (id == projectId)
```

Comment:
i don't love the implicit captures.  would prefer explicit (with 'ref' as-ap) to make it clear what state CanReuse is examining to determine this.  

That would also help make CanReuse not just a nested function in the middle of a loop (which reads very clnky to me).  The look is then just simple, and CanReuse is just liek a static helper, just scoped to this method to keep it away from otehr mtehods.

Reply:
Between the `ref dependencies` argument and the duplication of the already-confusing `projectId` and `id` values, making this local function static seemed worse than the current form.

Reply:
Going to agree with Cyrus that this does seem pretty clunky. Even if it was CanReuse(id) (and captured the rest) I think it'd almost be clearer, since it's at least then clear this is essentially a function that's going to give a different result for each id.

Reply:
Yeah if you capture *id* then it means you can move the local function out of the loop. It's more clear then that the captured stuff is just clearly ambient (why yes, there's only one dependency graph going on here) but the things that are changing are being explicitly passed in. Put another way, the function then behaves more as a function, in that the thing changing (the existing project Id) is the thing passed in, everything else is static.

Reply:
Actually you may want to disregard this and instead consider the method I'm proposing adding to ProjectDependencyGraph...

Reply:
Refactored according to https://github.com/dotnet/roslyn/pull/40358#discussion_r371563635